File: Program's_Contributed_By_Contributors/Python_Programs/rod cutting problem.py
# cut with max price result

def rodcutprob(w,wt,n,p):
    dp=[[-1 for x in range(w+1)] for x in range(n+1)]
    for i in range(n+1):
        for j in range(w+1):
            #print(i)
            if i==0:
                dp[i][j] = 0
            elif j == 0:
                dp[i][j] = 0
            elif wt[i-1] > j:
                dp[i][j] = dp[i-1][j] 
            else:
                #print(i,p[7])
                #print(p[i-1])#,i,j-wt[i-1],wt[i-1],j)
                                                        #change
                dp[i][j] = max(dp[i-1][j], p[i-1] + dp[i][j-wt[i-1]])
                
    return dp[n][w]

l=[1,2,3,4,5,6,7,8]
p=[1,5,8,9,10,17,17,20]
#w=8
#wt=[5,2,3]
#n=len(wt)
#n=5
n=8
w=8
wt=l
print(rodcutprob(w,wt,n,p))

File: Rod Cutting Problem.java
class Main
{
    // Function to find the best way to cut a rod of length `n`
    // where the rod of length `i` has a cost `price[i-1]`
    public static int rodCut(int[] price, int n)
    {
        // `T[i]` stores the maximum profit achieved from a rod of length `i`
        int[] T = new int[n + 1];
 
        // consider a rod of length `i`
        for (int i = 1; i <= n; i++)
        {
            // divide the rod of length `i` into two rods of length `j`
            // and `i-j` each and take maximum
            for (int j = 1; j <= i; j++) {
                T[i] = Integer.max(T[i], price[j - 1] + T[i - j]);
            }
        }
 
        // `T[n]` stores the maximum profit achieved from a rod of length `n`
        return T[n];
    }
 
    public static void main(String[] args)
    {
        int[] price = { 1, 5, 8, 9, 10, 17, 17, 20 };
        int n = 4;        // rod length
 
        System.out.print("Profit is " + rodCut(price, n));
    }
}
